### ISSUE TITLE
verifying what backend is being used by watcher

### ISSUE THREAD

**Issue by snoolord** [parcel-bundler/watcher#172]
Hi team,

I am considering switching from chokidar to watcher but need to be able to verify what backend watcher is using. It is not immediately clear to me how to verify that watcher is using "fs-events" or if it is using the fallback. 

Any help would be greatly appreciated! 🙇

Thanks

**Comment by bpasero** [comment]
You can explicitly pass the `backend` to use as `options`: https://github.com/parcel-bundler/watcher?tab=readme-ov-file#options

I am not aware that the watcher would fallback, I am not even sure what the fallback would be for macOS?

**Comment by snoolord** [comment]
thanks for the response @bpasero, I think the fallback will be 'brute-force', but I'm not sure how to verify that the backend is being used. Maybe there is an error message that watcher uses the fallback, but i
I want to build my app in production and have a log to verify. 

in chokidar, i do something like this
```
    chokidar.on('ready', () => {
      if (chokidar.platform === 'darwin' && chokidar.options.useFsEvents === false) {
        log(
          'The chokidar process falls back to polling. This may lead to a high CPU usage.'
        )
      } else if (chokidar.platform === 'darwin' && chokidar.options.useFsEvents === true) {
       log(
           The chokidar process utilizes fsevents. File changes are detected without polling.'
        )
      }
    })
    ```
    just wary as I have improperly bundled fs-events into my electron build and the production build chokidar was not using fs-events.

**Comment by bpasero** [comment]
I am quite certain that parcel watcher would use `fsevents` on macOS and `ReadDirectoryChangesW` on Windows. This module does not depend on the NPM module that you refer to at all.

**Comment by rogerweb** [comment]
It does. And it will use kqueue if FSEvents is not available. This is in the [READMEmd](https://olostec.olos.live/analytics/interaction-details/?interactionId=atendimento-oaa1-olostec-2024_03_20_11_54_11-104898), but the point is: how the application can tell in runtime which one the watcher is using?

**Comment by bpasero** [comment]
Ok, now I am curious as well. I was not aware that `fsevents` would not be available on macOS (unless maybe very old macOS versions). And looking at the sources its not very obvious to me how at all Kqueue would be used as fallback.

**Comment by mischnic** [comment]
This function gets called with either a concrete (set via options) backend or with `default`: https://github.com/parcel-bundler/watcher/blob/8dfd99440b8666e81cc6865f68697ab272dfe83d/src/Backend.cc#L26

All of these availability checks (apart from watchman) are really compile time checks, so it depends on what you compile for/against. Not really what is available at runtime (apart from watchman)

Fore example, I think it will only use kqueue by default on FreeBSD. And never on macOS because fsevents is always available there and takes priority.

**Comment by snoolord** [comment]
this may be only an issue when it comes to electron because it is all packaged. I ran into this issue outlined by the blog in https://www.hendrik-erz.de/post/electron-chokidar-and-native-nodejs-modules-a-horror-story-from-integration-hell. 
when electron is packaged, fs-events is not included and chokidar cannot use it. 
maybe parcel does not have this issue because it is compiled from C++. 
Nonetheless, it would be nice to be able to verify the backend at runtime. 
e.g. electron app installs watchman onto the user's system and uses parcel/watcher to watch the filesystem. would be nice to have a way to verify that it is indeed using watchman for linux/windows.

**Comment by mischnic** [comment]
Correct. Parcel doesn't use fsevents (the npm package), but fsevents (the C API).

**Comment by c5n8** [comment]
I have spent more time than I like to figure out how to install homebrew on linux on docker just to install watchman because they say it is a better backend, only to ask does it even matter?

**Comment by c5n8** [comment]
I found this in the log of `docker compose watch`

```
Syncing "directory" after changes were detected:
  - directory/.watchman-cookie-f8c2f2f04170-217-0
```

**Comment by c5n8** [comment]
I found this in the log of `docker compose watch`

```
Syncing "directory" after changes were detected:
  - directory/.watchman-cookie-f8c2f2f04170-217-0
```

Not the way that I like, but I can confidently say that watchman, when it is installed on Linux, will be used as backend.

Reference: [facebook.github.io/watchman/docs/cookies](https://facebook.github.io/watchman/docs/cookies)

**Comment by c5n8** [comment]
I found this in the log of `docker compose watch`

```
Syncing "directory" after changes were detected:
  - directory/.watchman-cookie-f8c2f2f04170-217-0
```

Not the way that I like, but this way we can verify that watchman is used as backend.

Reference: [facebook.github.io/watchman/docs/cookies](https://facebook.github.io/watchman/docs/cookies)
